### ISSUE TITLE
[Discussion] settings / configuration framework

### ISSUE THREAD

**Issue by yparitcher** [koreader/koreader#5470]
**TLDR: allow adjusting from anywhere, a setting that can be changed somewhere.**

Currently the setting are defined and changed in many places (depending on what type of setting it is).

To change a setting (add to the gesture manager etc) in many cases requires one to find where the setting is, and to find out what need to happen if the setting is changed (refresh screen etc).

This adds overhead and causes callbacks to be duplicated in the menu and elsewhere. (some settings are well defined with proper event handlers and therefore don't have this issue as much.)
the code to change many settings is defined in gesture manager (probably not the best place)

for example `inverse_reading_order` was added in 2016 #2075 but only shown in `ReaderPaging`. in august #5284 Frenzie added it to `ReaderPaging` and made it a global setting. However in order to change it in the gesture manager I had to add it again there in september. #5423

---

therefore i would like to implement some sort of framework (based on events?) to unify all settings that one can change somewhere in the UI.

this might be accomplished by having a dedicated function that is passed 
    a) setting name
    b) value to set (boolean, int, string, increment)

this function could then:
a) verify if the setting can be changed now (does a book need to be open), 
b) do any cleanup that changing the setting requires, 
c) or if the setting is well defined in its place just call the handler. 
d) it could also notify the user if the document needs to be reloaded.
e) it might return if the setting was changed successfully.

There might also be a function to get what values are correct for a setting (ie: list of font names, min/max value)

This would allow any part of koreader to attempt to change a setting without knowing how that setting is implemented.
This would greatly simplify much of the menu code, the gesture manager and would allow for easy profiles #5079 and other plugins

---

I would like to help work on this, it will probably be a lot of reworking existing code.

before I go anywhere I would like to hear what everyone has to say on this.
Do you have a better idea?
Does what i say make sense to you?
Do you think everything is fine the way it is?
Does this increase complexity with not enough gain?

The ideas i brought above are just a general outline of how i think it could be done, not a final say on my part.

in any case i would need help starting this off (where should this go? its own file? what should it be named? how should the function be called?)

**Comment by poire-z** [comment]
I'm usually conservative regarding such changes, so...

_Does what i say make sense to you?_ - Yes.
_Do you think everything is fine the way it is?_ - Yes.
_Does this increase complexity with not enough gain?_ - I think so.
_Do you have a better idea?_ - No, but I also don't see any problem :)

_the code to change many settings is defined in gesture manager (probably not the best place)_ - I think most of such code is defined in the module that brings the feature, and that's where it belongs, so you see the whole stuff involved (setting name, checks, loading and applying on book load, saving on change and book save...) in a single file. ReaderGesture is just a side thing (that we could do without), and I'm OK if some added complexity is put in it - rather than distributed everywhere.

_if the setting is well defined in its place just call the handler_ - I think that's the general quick solution to your problem.

_This would allow any part of koreader to attempt to change a setting without knowing how that setting is implemented._ - but in the end, you'd probably have some event broadcasted, because currently and with your proposal, you don't know what other part of the code need to react to that setting change (and as a broadcast costs a bit of CPU, you'd need to really see if it is needed, so to know if other parts of the code needs to react.)
And for 90% of our settings, there are quite localized, and don't need probagation. With your proposal, at minima, we'll get tons of things similar to eg. `onDictionaryOutputJustifyChange(justified=true|false)` (because why not include that in a profile! :) that would make the readerdictionary even more huge.

So, not surprising: not a fan :)
(well, at least until I see some thin/streamlined implementation that would not increase complexity - but hard to start thinking about it when I'm not hit by any problem :)

**Comment by Frenzie** [comment]
> Does this increase complexity with not enough gain?

I'm not quite sure if I understand what the advantage would be over just converting desired actions to events. All this stuff lived pretty much exclusively in the menu until earlier this year, when I put a concerted effort into doing exactly that. I think I got it up to a pretty decent coverage of the more important features.

NB I say actions because it might mean changing some display setting (and saving it for the next time) but it could also mean showing a widget.

> for example inverse_reading_order was added in 2016 #2075 but only shown in ReaderPaging. in august #5284 Frenzie added it to ReaderPaging and made it a global setting. However in order to change it in the gesture manager I had to add it again there in september. #5423

While there is a legacy problem of code duplication for a few select methods, it's a fairly small issue. A little helper method/module to make it easier to have a setting per-document or global without rewriting a bunch of the same boilerplate could be a good idea. It would help prevent potential future cases of #5284. I don't think that really needs to be some kind of "framework" though but feel free to make the proposal a bit more concrete in pseudocode/graphs/whatever. ;-)

> the code to change many settings is defined in gesture manager (probably not the best place)

The gesture manager would certainly be a candidate for some refactoring (i.e., split out a bunch of code into a separate module).

**Comment by yparitcher** [comment]
> I'm not quite sure if I understand what the advantage would be over just converting desired actions to events. All this stuff lived pretty much exclusively in the menu until earlier this year, when I put a concerted effort into doing exactly that. I think I got it up to a pretty decent coverage of the more important features.

Let me put it a little differently.
Any setting that has a event handler should be changed with that event, like is done now for most settings.

The definitions and handlers should stay local like is done now.

If one wants to change many settings together like is done in  ReaderGesture and also would be needed for profile support:

right now `ReaderGesture:gestureAction` is a function that takes a setting name and changes that setting. _if there is an event it uses that event_ 

To make a profile plugin would require duplicating most of that function.

If we were to split that out into a module it could be reused by profiles etc. 
this should not change that to toggle one setting you should still call the event directly.

Instead of having to call each event the ReaderGesture and profiles would just have to have a list of available settings and build a menu from them like `ReaderGesture:buildMenu` does now.

the list: https://github.com/koreader/koreader/blob/a771353aefb5d12372cb1a810c2cf08b37423928/frontend/apps/reader/modules/readergesture.lua#L655 might also be folded int this module so as not to be duplicated.

most settings are covered already, but if a new setting is made it would require less to add it everywhere (it would not be needed to add to the GM and profiles, just to this module)

**Comment by Frenzie** [comment]
> If we were to split that out into a module it could be reused by profiles etc.
this should not change that to toggle one setting you should still call the event directly.

Well that just sounds like a self-evident refactor. ;-)

**Comment by yparitcher** [comment]
fixed in  #6106
